File: packages/dart/lib/Seguranca/Ataques/AtaqueVetorDeterministico.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Com KEY e IV fixos, o keystream gerado precisa ser 100% determinístico -
/// chamar a função duas vezes com a mesma entrada tem que dar o mesmo
/// resultado, sempre. Se não der, alguma fonte de aleatoriedade vazou pra
/// dentro de uma função que deveria ser pura.
///
/// De brinde, esse ataque imprime o "vetor de referência" (Known Answer
/// Vector) pra esse par key/iv.
class AtaqueVetorDeterministico extends Ataque {
  @override
  String nome() => 'Determinismo (vetor de referência key/iv fixos)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chaveFixa = repetir('K', toBytes(alvo.chaveDeTeste()).length);
    final ivFixo = Uint8List(alvo.tamanhoIv());

    final ks1 = alvo.gerarKeystreamBruto(chaveFixa, ivFixo, 'enc', 32);
    final ks2 = alvo.gerarKeystreamBruto(chaveFixa, ivFixo, 'enc', 32);
    final ks3 = alvo.gerarKeystreamBruto(chaveFixa, ivFixo, 'enc', 32);

    final deterministico =
        _iguais(ks1, ks2) && _iguais(ks2, ks3);

    return ResultadoAtaque(
      nome(),
      !deterministico,
      deterministico ? Severidade.info : Severidade.critica,
      deterministico
          ? 'Determinístico em 3 chamadas. Vetor de referência '
                '(key=64x"K", iv=zeros, prop=enc, 32 bytes): '
                '${bytesToHex(ks1)}'
          : 'NÃO determinístico! 3 chamadas com a mesma entrada deram '
                'resultados diferentes: ${bytesToHex(ks1)} / '
                '${bytesToHex(ks2)} / ${bytesToHex(ks3)}',
      {'vetor_hex': bytesToHex(ks1)},
    );
  }

  bool _iguais(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
}

File: packages/julia/src/Seguranca/Ataques/AtaqueVetorDeterministico.jl
"""
Com KEY e IV fixos, o keystream gerado precisa ser 100% determinístico.
Também imprime o vetor de referência (KAT) para esse par key/iv.
"""
struct AtaqueVetorDeterministico <: Ataque end

nome(::AtaqueVetorDeterministico) = "Determinismo (vetor de referência key/iv fixos)"

function executar(a::AtaqueVetorDeterministico, alvo::AlvoCriptografico)
    chave_fixa = repeat("K", length(codeunits(chave_de_teste(alvo))))
    iv_fixo = zeros(UInt8, tamanho_iv(alvo))

    ks1 = gerar_keystream_bruto(alvo, chave_fixa, iv_fixo, "enc", 32)
    ks2 = gerar_keystream_bruto(alvo, chave_fixa, iv_fixo, "enc", 32)
    ks3 = gerar_keystream_bruto(alvo, chave_fixa, iv_fixo, "enc", 32)

    deterministico = ks1 == ks2 && ks2 == ks3
    hex1 = bytes2hex(ks1)

    detalhes = deterministico ?
               "Determinístico em 3 chamadas. Vetor de referência (key=64x\"K\", iv=zeros, prop=enc, 32 bytes): " *
               hex1 :
               "NÃO determinístico! 3 chamadas com a mesma entrada deram resultados diferentes: " *
               hex1 * " / " * bytes2hex(ks2) * " / " * bytes2hex(ks3)

    return ResultadoAtaque(
        nome(a),
        !deterministico,
        deterministico ? "info" : "critica",
        detalhes,
        Dict{String,Any}("vetor_hex" => hex1),
    )
end
